sled-agent/src/updates.rs

```diff
@@ -0,0 +1,136 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+//! Management of per-sled updates
+
+use schemars::JsonSchema;
+use serde::Deserialize;
+use std::path::Path;
+
+#[cfg(test)]
+use crate::mocks::MockNexusClient as NexusClient;
+#[cfg(not(test))]
+use nexus_client::Client as NexusClient;
+
+#[derive(thiserror::Error, Debug)]
+pub enum Error {
+    #[error("I/O Error: {0}")]
+    Io(#[from] std::io::Error),
+
+    #[error("Failed to contact nexus: {0}")]
+    Nexus(anyhow::Error),
+
+    #[error("Failed to read response from Nexus: {0}")]
+    Response(reqwest::Error),
+}
+
+// TODO: De-duplicate this struct with the one in iliana's PR?
+//
+// This should likely be a wrapper around that type.
+#[derive(Clone, Debug, Deserialize, JsonSchema)]
+pub struct UpdateArtifact {
+    pub name: String,
+    pub version: i64,
+    pub kind: UpdateArtifactKind,
+}
+
+// TODO: De-dup me too.
+#[derive(Clone, Debug, Deserialize, JsonSchema)]
+pub enum UpdateArtifactKind {
+    Zone,
+}
+
+impl UpdateArtifact {
+    fn artifact_directory(&self) -> &'static Path {
+        match self.kind {
+            UpdateArtifactKind::Zone => Path::new("/var/tmp/zones"),
+        }
+    }
+
+    /// Downloads an update artifact.
+    ///
+    /// The artifact is eventually stored in the path:
+    ///     <artifact_directory()> / <artifact name>
+    ///
+    /// Such as:
+    ///     /var/tmp/zones/myzone
+    ///
+    /// While being downloaded, it is stored in a path also containing the
+    /// version:
+    ///     <artifact_directory()> / <artifact name> - <version>
+    ///
+    /// Such as:
+    ///     /var/tmp/zones/myzone-3
+    pub async fn download(&self, nexus: &NexusClient) -> Result<(), Error> {
+        let file_name = format!("{}-{}", self.name, self.version);
+        let response = nexus
+            .cpapi_artifact_download(&file_name)
+            .await
+            .map_err(|e| Error::Nexus(e))?;
+
+        let mut path = self.artifact_directory().to_path_buf();
+        tokio::fs::create_dir_all(&path).await?;
+
+        // We download the file to a location named "<artifact-name>-<version>".
+        // We then rename it to "<artifact-name>" after it has successfully
+        // downloaded, to signify that it is ready for usage.
```

Comment:
this strikes me as odd -- should future parts of sled-agent be aware of what artifact version it's trying to apply? i feel like there's a possible race here where nexus tells the sled agent to apply version N+1 of an artifact in the middle of it applying version N

Reply:
At the end of the day, the sled agent needs to set "something" as active, right? Whether that's "flash the firmware", "write the OS partition", or "update a file".

You have a good point - we should probably add concurrency control around the updates to make them safe amid concurrent requests - but that's kinda what this renaming silliness is all about:

1. Download to a "unique" spot. I dunno if the version aspect really matters here, I just wanna make sure stream all necessary data to the sled before using it.
2. "Apply" the update atomically